BiometricLoginKotlin/app/src/main/java/com/example/biometricloginsample/LoginActivity.kt

```diff
@@ -0,0 +1,194 @@
+            val secretKeyName = getString(R.string.secret_key_name)
+            val cipher = cryptographyManager.getInitializedCipherForDecryption(
+                secretKeyName, encryptedServerTokenWrapper.initializationVector
+            )
+            biometricPrompt =
+                BiometricPromptUtils.createBiometricPrompt(this, ::decryptServerTokenFromStorage)
+            val promptInfo = BiometricPromptUtils.createPromptInfo(this)
+            biometricPrompt.authenticate(promptInfo, BiometricPrompt.CryptoObject(cipher))
+        }
+    }
+
+    private fun decryptServerTokenFromStorage(authResult: BiometricPrompt.AuthenticationResult) {
+        authResult?.cryptoObject?.cipher?.apply {
+            val plaintext =
+                cryptographyManager.decryptData(encryptedServerTokenWrapper.ciphertext, this)
+            SampleAppUser.fakeToken = plaintext
+            // now that you have the token, you can query server for everything else
+            // the only reason we call this fakeToken is because we didn't really get it from
+            // the server. In your case, you will have gotten it from the server the first time
+            // and therefore, it's a real token.
+
+            updateApp(getString(R.string.already_signedin))
+        }
+    }
+
+    // USERNAME + PASSWORD SECTION
+
+    private fun loginWithPasswordManager() {
+        val loginWithPasswordViewModel by viewModels<LoginWithPasswordViewModel>()
+        loginWithPasswordViewModel.loginWithPasswordFormState.observe(this, Observer {
+            val loginState = it ?: return@Observer
+            when (loginState) {
+                is SuccessfulLoginFormState -> binding.login.isEnabled = loginState.isDataValid
+                is FailedLoginFormState -> {
+                    loginState.usernameError?.let { binding.username.error = getString(it) }
+                    loginState.passwordError?.let { binding.password.error = getString(it) }
+                }
+            }
+        })
+
+        binding.username.afterTextChanged {
+            loginWithPasswordViewModel.onLoginDataChanged(
+                binding.useBiometrics.text.toString(),
+                binding.password.text.toString()
+            )
+        }
+
+        binding.password.apply {
+            afterTextChanged {
+                loginWithPasswordViewModel.onLoginDataChanged(
+                    binding.username.text.toString(),
+                    binding.password.text.toString()
+                )
+            }
+
+            setOnEditorActionListener { _, actionId, _ ->
+                when (actionId) {
+                    EditorInfo.IME_ACTION_DONE ->
+                        loginWithPassword(
+                            binding.username.text.toString(),
+                            binding.password.text.toString()
+                        )
+                }
+                false
+            }
+        }
+        binding.login.setOnClickListener {
```

Comment:
This is not the right place to set click listeners.

Reply:
Actually this is directly from the Android Studio Login Activity. I am taking a typical login recommendation and showing devs  how to add Biometrics to it.

Reply:
This is _not_ the way it's done in the template. In the template, all of this code is in `onCreate` which matches with what I commented above.

Reply:
Unresolving as the comment's are not resolved.

Reply:
see latest commit